### ISSUE TITLE
libxml_disable_entity_loader status must be restored to previous value

### ISSUE THREAD

**Issue by sgomez** [simplesamlphp/simplesamlphp#241]
At the end of _/www/_include.php_ we can found this:

```php
// disable XML external entity loading explicitly
libxml_disable_entity_loader();
```

But this function has a bug (or a feature, I don't know): https://bugs.php.net/bug.php?id=64938
The configuration is shared between threads, so when this function is called, the entity loader is disabled for all php threads forever (or until php-fpm restarts). The problem is the default value is false and not true, so it could cause problems with  applications than interact with simplesamlphp and expects the default value (Symfony and FOSUserBundle in my case).

The polite way should be deactivate and restore the previous value every time than we process a XML file:

```php
$previous = libxml_disable_entity_loader(true);
// Some XML stuff
libxml_disable_entity_loader($previous)
```

But I guess that this requires some work. I'd like to help but I don't want to work in this without know if the push will be accepted in the first place.

**Comment by thijskh** [comment]
If we can do it more targeted I'm all for it but the core reason while we call this function is that we want to be sure that nowhere these remote entities are loaded.

I see that the bug you referenced has been fixed in PHP 5.5 and up which should limit the trouble somewhat that this causes.

I'm a bit puzzled though that Symfony and FOSUserBundle would require to be able to expand remote XML entities which is in my opinion a rather obscure (and dangerous) aspect of XML processing.

**Comment by sgomez** [comment]
I saw that this bug is closed but I am working with php 5.6.12 and I have the same issue with php-fpm in OSX and Ubuntu.

So I keep searching and I found that there are another related bug: https://bugs.php.net/bug.php?id=62577. FOSUserBundle is using _simplexml_load_file()_ to read local files and _libxml_disable_entity_loader()_ is blocking local files too. 

But the point is that Symfony, MediaWiki, Guzzle and some others are restoring the previous state and I think that SimpleSAMLPHP shouldn't modify the default PHP behaviour. 

Some references:
- [MediaWiki](https://www.mediawiki.org/wiki/XML_External_Entity_Processing)
- [PHP Security](http://phpsecurity.readthedocs.org/en/latest/_articles/PHP-Security-Default-Vulnerabilities-Security-Omissions-And-Framing-Programmers.html#xml-injection-attacks)

**Comment by thijskh** [comment]
Yes, the current blanket override is not ideal. I think a change where the entity_loader is disabled is indeed better but the challenge would be to guarantee it always happens.

**Comment by sgomez** [comment]
The fastest way could be something like that in _www/_include.php_:

```php
// disable XML external entity loading explicitly
$previous = libxml_disable_entity_loader(true);
// and restore XML external entity loading on exit
register_shutdown_function('libxml_disable_entity_loader', $previous);
```

It works like a charm.

**Comment by thijskh** [comment]
Seems like a straightforward solution to me.

**Comment by jaimeperez** [comment]
Hi guys!

If I recall correctly, we decided to disable the XML entity loader to prevent XXE attacks with some PHP versions, but since we are requiring PHP 5.3 now, that was no longer an issue, right? If that's the case, why don't just removing the call to `libxml_disable_entity_loader()`?

Maybe I'm completely wrong and it actually depends on how your PHP version was compiled. If that's the case, then the last suggestion by Sergio makes perfect sense :+1:

**Comment by thijskh** [comment]
I'm not aware that PHP 5.3 is better in this regard than its predecessors. There was some mention that PHP 5.3.23 disables it but I cannot confirm that (it looks like it only disables it for SOAP). I think it's still needed.

**Comment by thijskh** [comment]
Some tests confirm that an XXE poc does not work on any of Debian squeeze/wheezy/jessie/stretch, but it is effective on RHEL 6 (PHP 5.3.3). RHEL 7 seems to have issued an erratum that fixes it. So it's OS (libxml2) dependant whether you're safe by default. I'd prefer to keep the defense in depth if possible.

**Comment by jaimeperez** [comment]
Thanks for checking Thijs! I agree then, we should keep this, but in a way that doesn't disrupt other applications.

Sergio, would you mind providing a PR with the change?

**Comment by sgomez** [comment]
From libxml2 >= 2.9.0, the entity loader is disabled by default. But anyway, the problem is that on PHP the entity loader is enabled by default (some BC question maybe?). So the entity loader must be disabled, that is not under discussion.

But IMHO, disabling the entity loader without checking every time we read a XML document don't resolve the issue. Any third party module could enable it again, lulling us into a false sense of security. So, my last proposal (to use `register_shutdown_function()`) is not a solution.

So, I think the best way is disabling and restoring the entity loader before any time SimpleSAML use the vulnerable functions (just like the two documents I linked before describe):

* `XMLReader::read()`
* `DOMDocument::loadXML()`
* `DOMDocument::loadHTML()`
* `simplexml_load_string()`

That's the only way than we can be sure we are safe from XXE attacks. I just saw only `DOMDocument::loadXML()` is used. So, if you feel ok with this I'd prefer to do a PR protecting the vulnerable functions instead of a global disabling.

**Comment by thijskh** [comment]
Yes, that seems like a good approach.

**Comment by jaimeperez** [comment]
I have a couple of concerns:

* What if the bug about this function not being thread-safe is present in your PHP installation? Of course, moving the call to `libxml_disable_entity_loader()` closer to where it is actually relevant makes it less likely that an incident occurs due to concurrency, but it would still be possible.

* Additionally, and this is the biggest problem I see, it would be a **huge** amount of work to protect all calls to `loadXML()` like this. We would need to create an auxiliary function that disables entity loading, loads a DOMDocument from XML, then restores previous status, and substitute all calls to `DOMDocument::loadXML()` by the new auxiliary function.

If the concern is third-party modules, then what they do or not is their responsibility. We can do however everything at hand to avoid having a problem due to a module misbehaving, but that doesn't necessarily imply that we control this so fine-grained. We know where we parse XML documents (metadata, requests, responses) in the code, so we could just surround the outermost pieces of code with the calls to `libxml_disable_entity_loader()`, as no modules should be executed then. Concurrency with misbehaving modules could be a problem then, of course, but just as with any other PHP software running in the same server.

So honestly I don't know. There's no way to be completely sure that everything is going to work fine. We can reach a level of protection high enough, but that means a lot of work. If you feel like you can provide a PR with the auxiliary function and all the calls to `loadXML()` replaced, then I'm fine with that.

**Comment by sgomez** [comment]
> If you feel like you can provide a PR with the auxiliary function and all the calls to loadXML() replaced, then I'm fine with that.

No problem. I'm on it.

A final question: The _simplesaml/saml2_ package already have two auxiliary functions for that: ```SAML2_DOMDocumentFactory::fromString()``` and ```SAML2_DOMDocumentFactory::fromFile()``` and they must be protected too. So, I'm going to start to fix them (with test cases). They could be used like the auxiliary functions that you are talking about. What do you think?

**Comment by jaimeperez** [comment]
Sounds like a plan! :+1:

**Comment by sgomez** [comment]
Now than SAML2_DOMDocumentFactory is safe to XXE attacks, I refactorized all the methods that are using ```DOMDocument::loadXML()``` for  ```SAML2_DOMDocumentFactory::fromString()```. See PR #271.

But this saml2 class appears on 1.6.1. So I changed the composer.json to require this version at least. Anyway, this bug is not resolved until a new simplesamlphp/saml2 version be released with PR simplesamlphp/saml2#52 on it.
